libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs

```diff
@@ -85,88 +91,106 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
+        /// <summary>
+        /// Gets a "normalized" title for a <see cref="Choice"/> object.
+        /// </summary>
+        /// <param name="choice">The choice object.</param>
+        /// <returns>The normalized title.</returns>
+        /// <exception cref="InvalidOperationException">The choice does not have a valid action title or value to use as a title.</exception>
+        private static string GetTitle(Choice choice)
+        {
+            if (choice is null) return null;
+
+            if (!string.IsNullOrWhiteSpace(choice.Action?.Title)) return choice.Action.Title.Trim();
+            if (!string.IsNullOrWhiteSpace(choice.Value)) return choice.Value.Trim();
+
+            throw new InvalidOperationException("Each choice must specify either an action title or a value.");
+        }
+
+        public static Activity List(IEnumerable<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
         {
             return List(ToChoices(choices), text, speak, options);
         }
 
-        public static Activity List(List<Choice> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        public static Activity List(IEnumerable<Choice> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
         {
-            choices = choices ?? new List<Choice>();
             options = options ?? new ChoiceFactoryOptions();
 
             bool includeNumbers = options.IncludeNumbers ?? true;
 
             // Format list of choices
-            var connector = string.Empty;
-            var txt = (text ?? string.Empty);
-            txt += "\n\n   ";
-
-            for (var index = 0; index < choices.Count; index++)
+            var separator = string.Empty;
+            var sb = new StringBuilder();
+            if (text != null)
             {
-                var choice = choices[index];
-
-                var title = choice.Action != null && choice.Action.Title != null ? choice.Action.Title : choice.Value;
+                sb.Append(text);
+            }
+            sb.Append(" ");
 
-                txt += connector;
-                if (includeNumbers)
-                {
-                    txt += (index + 1).ToString() + ". ";
-                }
-                else
+            if (choices != null)
+            {
+                int count = choices.Where(c => c != null).Count();
+                int index = 1;
+                foreach (var choice in choices.Where(c => c != null))
                 {
-                    txt += "- ";
+                    string title = GetTitle(choice);
+                    sb.Append(separator);
+                    sb.Append((includeNumbers) ? $"{index}. " : "- ");
+                    sb.Append(title);
+
+                    separator = Environment.NewLine + " ";
```

Comment:
This ends up generating a bunch of garbage `string`s (due to the concatenation) on each pass through the loop. Suggest moving the `Environment.NewLine + " "` into a static to avoid this _OR_ a second local variable _OR_, wrap this with a check that will only assign the separator to the new string on the first pass when it's `Length == 0` (`string.Empty` check).

Personally, for this pattern, what I've usually done in the past is _ALWAYS_ append the separator _AFTER_ the value and then outside the for loop you can back up the `StringBuffer` to truncate those characters off for the last value. You'd do that like this after the loop:

```CSharp
sb.Length -= separator.Length;
```

Make sense?